generation/new_client_hermetic_build/new-client.py

```diff
@@ -0,0 +1,332 @@
+        "GAPICs": [{"proto_path": p} for p in proto_paths],
+    }
+
+    __add_item_if_set(new_library, "requires_billing", requires_billing)
+    __add_item_if_set(new_library, "rest_documentation", rest_docs)
+    __add_item_if_set(new_library, "rpc_documentation", rpc_docs)
+    __add_item_if_set(new_library, "distribution_name", distribution_name)
+    __add_item_if_set(new_library, "api_reference", api_reference)
+    __add_item_if_set(new_library, "codeowner_team", codeowner_team)
+    __add_item_if_set(new_library, "excluded_dependencies", excluded_dependencies)
+    __add_item_if_set(new_library, "excluded_poms", excluded_poms)
+    __add_item_if_set(new_library, "googleapis_commitish", googleapis_committish)
+    __add_item_if_set(new_library, "issue_tracker", issue_tracker)
+    __add_item_if_set(new_library, "extra_versioned_modules", extra_versioned_modules)
+
+    config["libraries"].append(new_library)
+    config["libraries"] = sorted(config["libraries"], key=__compute_library_name)
+
+    with open(path_to_yaml, "w") as file_stream:
+        yaml.dump(config, file_stream)
+
+
+def __add_item_if_set(target: dict, key: str, value: any) -> None:
+    if value is not None:
+        target[key] = value
+
+
+def __compute_library_name(library: dict) -> str:
+    if "library_name" in library:
+        return f'java-{library["library_name"]}'
+    return f'java-{library["api_shortname"]}'
+
+
+def __get_proto_paths(proto_path: str, committish: str) -> List[str]:
+    version_re = re.compile(r"v\d[\w\d]*")
+    is_library_version = lambda p: version_re.match(p.split("/")[-1]) is not None
+
+    if is_library_version(proto_path):
+        raise ValueError(
+            "Versioned proto_paths are not supported, please provide the root"
+            " proto_path. For example `google/datastore` instead of `google/datastore/v1`."
+            " If all you need is to add a new version to an existing library,"
+            " you need to manually add the new version to `generation_config.yaml`"
+        )
+
+    # if the proto path is not versioned, we must search for all versions
+    # in googleapis
+    if os.path.isdir("./googleapis"):
+        googleapis_repo = Repo("./googleapis")
+    else:
+        googleapis_repo = Repo.clone_from(
+            "https://github.com/googleapis/googleapis", "./googleapis"
+        )
+    googleapis_repo.git.checkout(committish)
+    subdirs = os.walk(f"./googleapis/{proto_path}")
+    proto_paths = []
+    for subdir_path, _, _ in subdirs:
+        if is_library_version(subdir_path) and __is_releasable_version(subdir_path):
+            print(subdir_path)
+            proto_paths.append(subdir_path.replace("./googleapis/", ""))
+    shutil.rmtree("./googleapis")
+    return proto_paths
+
+
+def __is_releasable_version(versioned_proto_path):
+    """true if the library definition found in `versioned_proto_path`:
+    - has a BUILD file
+    - the BUILD file has a `java_gapic_assembly_gradle_pkg` rule
```

Comment:
Is there any use case that the BUILD file does not have a `java_gapic_assembly_gradle_pkg` rule?

Reply:
This is the heuristic we determined with @JoeWang1127 to tell if a library version has been released. @JoeWang1127 Would you mind adding a link to some doc or source to validate this?

Reply:
For example, https://github.com/googleapis/googleapis/tree/master/google/bigtable/v1 has no BUILD file, but I'm still looking for one that doesn't have a `java_gapic_assembly_gradle_pkg` rule.

Reply:
I'm removing this logic in favor of supporting versioned proto_paths only

Reply:
I remembered I have seen an example of this, however I can't find it now.

We can ignore this case and revise it later.